VSR: `send_request_blocks()` more proactively

sentientwaffle opened the pull request:
(I recommend reviewing commit-by-commit.)

## Bug

This is primarily important during the initial phase of state sync, when we sync the client sessions and manifest trailers.

Requesting grid blocks is triggered by receiving grid blocks -- when we receive a block that we need, we restore budget and if we have enough budget we make a new request.

This works great during table sync, but during trailer sync it doesn't work well. The trailers are repaired sequentially (free set → client sessions → manifest) but between each of these we need to do a read (to check for the grid block on our disk). That means that e.g. at the moment we repair the last free set block, we restore budget, then check whether we have more blocks to repair, but we don't yet. In a moment the client sessions read finishes (fails) and we queue a block for repair, but it is too late. So we wait for `grid_repair_message_timeout` (~500ms) before triggering the next request. Then this happens again during the transition from client sessions to manifest. This adds about a second of latency to state sync!

## Fix

To fix, we now send the block request from `repair()`. `repair()` is on a more frequent timer, but also it is called proactively often.
The `grid_repair_message_timeout` is still used to restore budget. (Renamed to `grid_repair_budget_timeout`).

**Amendment:** See https://github.com/tigerbeetle/tigerbeetle/pull/3185/commits/8f7dcbfe10e3d854252a68176c8ae8e0598a47ee. We still request blocks from `grid_repair_budget_timeout` after all, but it is not the main way blocks are repaired.

Some other possible fixes:
1. Concurrently repair all the trailers.
2. Send a summary of trailer blocks in the SV.
3. Add a new "survey" block which includes the address/checksums of all trailer blocks.
4. Wire up an event from grid → replica when a block is queued for repair.
5. Reduce `grid_repair_message_timeout`.

Probably we should eventually implement 1 and 2-or-3, but I think this change is useful even then, so that we are a little more proactive about repair.

I wanted to avoid 4, to avoid coupling replica + grid together in that direction.
5 only reduces the scale of the problem, but doesn't actually solve it.

## Repair Budget

Track in-flight block requests via `RepairBudgetGrid`.

Without in-flight request tracking, if a replica requests the same block multiple times in the same budget interval, then when the block arrives, the budget is still only incremented once, leaving the budget starved for the remainder of the interval. This is a very common case when there are few blocks being requested, such as during trailer sync.

As part of this, lift `request_blocks` construction into replica.

## Logs

Logs of a replica state syncing trailers before and after this change is applied (with some unrelated logs elided).

Mainly pay attention to the timestamps.

<details>
<summary>Before</summary>

```
2025-08-15 17:09:35.718Z info(replica): 0n: sync: ops=470432..487711
2025-08-15 17:09:36.199Z info(replica): 0n: on_block: fulfilled address=1794 checksum=222976420026131580906061225735420155686 free_set
2025-08-15 17:09:36.199Z info(replica): 0n: on_block: repairing address=1794 checksum=222976420026131580906061225735420155686 free_set
2025-08-15 17:09:36.202Z info(replica): 0n: on_block: fulfilled address=1795 checksum=58593412436848580214907331841939497839 free_set
2025-08-15 17:09:36.202Z info(replica): 0n: on_block: repairing address=1795 checksum=58593412436848580214907331841939497839 free_set
2025-08-15 17:09:36.701Z info(replica): 0n: on_block: fulfilled address=1725 checksum=212153718505234864535153087633500583820 client_sessions
2025-08-15 17:09:36.701Z info(replica): 0n: on_block: repairing address=1725 checksum=212153718505234864535153087633500583820 client_sessions
2025-08-15 17:09:37.207Z info(replica): 0n: sync: 35 tables (by level: { 35, 0, 0, 0, 0, 0, 0 })
```

</details>

<details>
<summary>After</summary>

```
2025-08-15 17:28:10.421Z info(replica): 0n: sync: ops=222752..224671
2025-08-15 17:28:10.422Z info(replica): 0n: on_block: fulfilled address=939 checksum=73337053751425680451193589040484341376 free_set
2025-08-15 17:28:10.422Z info(replica): 0n: on_block: repairing address=939 checksum=73337053751425680451193589040484341376 free_set
2025-08-15 17:28:10.425Z info(replica): 0n: on_block: fulfilled address=936 checksum=149582764964113425824743432799185874373 free_set
2025-08-15 17:28:10.425Z info(replica): 0n: on_block: repairing address=936 checksum=149582764964113425824743432799185874373 free_set
2025-08-15 17:28:10.427Z info(replica): 0n: on_block: fulfilled address=883 checksum=28498878219503029859547433952126740385 client_sessions
2025-08-15 17:28:10.427Z info(replica): 0n: on_block: repairing address=883 checksum=28498878219503029859547433952126740385 client_sessions
2025-08-15 17:28:10.430Z info(replica): 0n: sync: 21 tables (by level: { 21, 0, 0, 0, 0, 0, 0 })
```

</details>

The budget tracking change improves table sync performance as well, but that is higher variance between runs so it is hard to get a decent measurement of how much it actually improved.

sentientwaffle committed:
VSR: `send_request_blocks()` via `repair()`

This is primarily important during the initial phase of state sync, when we sync the client sessions and manifest trailers.

Requesting grid blocks is triggered by receiving grid blocks -- when we receive a block that we need, we restore budget and if we have enough budget we make a new request.

This works great during table sync, but during trailer sync it doesn't work well. The trailers are repaired sequentially (free set → client sessions → manifest) but between each of these we need to do a read (to check for the grid block on our disk). That means that e.g. at the moment we repair the last free set block, we restore budget, then check whether we have more blocks to repair, but we don't yet. In a moment the client sessions read finishes (fails) and we queue a block for repair, but it is too late. So we wait for `grid_repair_message_timeout` (~500ms) before triggering the next request. Then this happens again during the transition from client sessions to manifest. This adds about a second of latency to state sync!

To fix, we now send the block request from `repair()`. `repair()` is on a more frequent timer, but also it is called proactively often.
The `grid_repair_message_timeout` is still used to restore budget.

Some other possible fixes:
1. Concurrently repair all the trailers.
2. Send a summary of trailer blocks in the SV.
3. Add a new "survey" block which includes the address/checksums of all trailer blocks.
4. Wire up an event from grid → replica when a block is queued for repair.
5. Reduce `grid_repair_message_timeout`.

Probably we should eventually implement 1 and 2-or-3, but I think this change is useful even then, so that we are a little more proactive about repair.

I wanted to avoid 4, to avoid coupling replica + grid together in that direction.
5 only reduces the scale of the problem, but doesn't actually solve it.

sentientwaffle committed:
VSR: Rename grid_repair_{message → budget}_timeout

Since it doesn't send a message anymore.

sentientwaffle committed:
VSR: Lift `request_blocks` construction into replica

This is setup for the following commit, which will track in-flight block requests as part of the `RepairBudgetGrid`.

But also, this code was previously part of `Grid`/`GridBlocksMissing` because it manipulated their state internally as a workaround for the lack of proper in-flight request checking.

`GridBlocksMissing` still has `faulty_blocks_repair_index` but this is just an optimization. In some cases we may still need to iterate the whole `faulty_blocks` to find a block to request, since removals are via `swapRemoveAt`.

sentientwaffle committed:
VSR: Track in-flight block requests via RepairBudgetGrid

Without in-flight request tracking, if a replica requests the same block multiple times in the same budget interval, then when the block arrives, the budget is still only incremented once, leaving the budget starved for the remainder of the interval. This is a very common case when there are few blocks being requested, such as during trailer sync.

sentientwaffle committed:
VOPR: Request blocks even on view-changing backup

Request blocks even on view-changing backup. This is to avoid a "block found in core" panic.

Seed: ./zig/zig build -Drelease vopr -- 4961535418630898736

matklad commented:
Needs a rebase!

matklad commented:
```
/home/runner/work/tigerbeetle/tigerbeetle/src/vsr/replica.zig:10853:25: error: no field named 'grid_repair_message_timeout' in struct 'vsr.replica.ReplicaType(state_machine.StateMachineType(storage.StorageType(io.linux.IO),.{ .release = .{ ... }, .message_body_size_max = 1048320, .lsm_compaction_ops = 32 }),message_bus.MessageBusType(.replica),storage.StorageType(io.linux.IO),aof.AOFType(io.linux.IO))'
            assert(self.grid_repair_message_timeout.ticking);
```

sentientwaffle commented:
Ah, sorry :facepalm:

matklad commented on src/vsr/replica.zig:
```diff
@@ -10832,11 +10850,13 @@ pub fn ReplicaType(
             maybe(self.state_machine_opened);
-
-            self.grid_repair_message_timeout.reset();
+            if (!self.solo()) {
```
Why isn't this `assert(!self.solo())`? Would it be ok if 

* self.solo() istrue
* assert(self.repair_messages_budget_grid.available >=           constants.grid_repair_request_max); is false
* and we call self.repair_messages_budget_grid.decrement
?

sentientwaffle replied:
Not sure I understand your suggestion:

> Why isn't this assert(!self.solo())?

Right now we call `send_request_blocks` even for solo replicas, its just that those replicas shouldn't ever need to request anything. If they did, `choose_any_other_replica` would panic.

But possibly before that -- this function assumes that _if_ something is being requested, we have at least `grid_repair_request_max` budget. The case where this might be false is for solo replicas, since their `repair_messages_budget_grid_refill` is lower (it is scaled by replica count).

Note that `decrement` actually asserts `budget.available > 0`.

matklad replied:
Ah, got it, thank! I was worried that the `decrement` call is correct only because we guard it by _this_ assert, but _this_ assert doesn't work for `.solo`. But solos wouldn't ever decrement, so that's fine! Though, maybe assert

```
            if (requests_count == 0) return;
            assert(!self.solo());
```

matklad commented on src/vsr/replica.zig:
```diff
@@ -10847,13 +10867,52 @@ pub fn ReplicaType(
+                        }
+                    }
+                } else break @intCast(i);
+            } else @intCast(requests.len);
```
Complex control flow! I think it can be simplified if we don't try to intertwine loop over request buffer and two loops over request sources, and just use an index? 

```zig
            const requests = std.mem.bytesAsSlice(
                vsr.BlockRequest,
                message.buffer[@sizeOf(Header)..],
            )[0..constants.grid_repair_request_max];
            assert(requests.len > 0);
            var requests_count: u32 = 0;

            // Prioritize requests for blocks with stalled Grid reads, so that commit/compaction can
            // continue. We divide the buffer up between `read_global_queue` and
            // `blocks_missing.faulty_blocks` so that we always request blocks from both queues.
            // Surplus from `blocks_missing.faulty_blocks` may be used by `read_global_queue`.
            const request_faults_count_max = requests.len -
                @min(@divFloor(requests.len, 2), self.grid.blocks_missing.faulty_blocks.count());
            assert(request_faults_count_max > 0);
            assert(request_faults_count_max <= requests.len);
            assert(request_faults_count_max >= @divFloor(requests.len, 2));

            var grid_faults = self.grid.read_global_queue.iterate();
            while (grid_faults.next()) |read_fault| {
                if (requests_count >= request_faults_count_max) break;

                if (self.repair_messages_budget_grid.decrement(.{
                    .address = read_fault.address,
                    .checksum = read_fault.checksum,
                })) {
                    requests[requests_count] = .{
                        .block_address = read_fault.address,
                        .block_checksum = read_fault.checksum,
                    };
                    requests_count += 1;
                }
            }

            for (0..self.grid.blocks_missing.faulty_blocks.count()) |_| {
                if (requests_count >= requests.len) break;

                if (self.grid.blocks_missing.next_request()) |missing_request| {
                    if (self.repair_messages_budget_grid.decrement(.{
                        .address = missing_request.block_address,
                        .checksum = missing_request.block_checksum,
                    })) {
                        requests[requests_count] = missing_request;
                        requests_count += 1;
                    }
                }
            }
```

If we go with this solution ,recommend `request_buffer/request_count` naming scheme!

sentientwaffle replied:
Thanks, this is much clearer!

matklad commented on src/vsr.zig:
```diff
@@ -1843,33 +1843,67 @@ pub const RepairBudgetGrid = struct {
+        budget.requested.deinit(gpa);
+    }
+
+    fn assert_internally_consistent(budget: *RepairBudgetGrid) void {
```
Maybe just call this `fn invariants()` like we do for replica? Or `assert_invariants`

(I know that we use this name a bunch, so maybe we should change all)

sentientwaffle replied:
I like `assert_invariants` :+1:

matklad commented on src/vsr.zig:
```diff
@@ -1843,33 +1843,67 @@ pub const RepairBudgetGrid = struct {
+    pub fn decrement(budget: *RepairBudgetGrid, block_identifier: BlockIdentifier) bool {
+        budget.assert_internally_consistent();
+        defer budget.assert_internally_consistent();
+        assert(budget.available > 0);
+        assert(block_identifier.address > 0);
+
+        const gop = budget.requested.getOrPutAssumeCapacity(block_identifier);
+        if (gop.found_existing) {
+            return false;
+        } else {
+            budget.available -= 1;
+            return true;
+        }
```
TBH, it feels weird that the contact here is that the budget must be > 0, but we return false. It feels that "already repairing" and "out of budget" are maybe the same, as far as the caller is concerned? But maybe not. For blocks, we indeed don't want to drip feed one block per request, so we are asserting that we have enough budget available. 

Ok. convinced myself that it works as expected!

sentientwaffle committed:
VSR: Rename assert_internally_consistent → assert_invariants

matklad commented on src/vsr.zig:
```diff
@@ -1843,33 +1843,67 @@ pub const RepairBudgetGrid = struct {
+        }
     }
 
     pub fn refill(budget: *RepairBudgetGrid, amount: u32) void {
```
Feels weird that we allow the caller to control amount, but then we always clear requested. Maybe `increment` should alway refill by one, and `refill` should always refill max amount? 

We currently allow `budget.refil(1)`, but with cleaning logic it doesn't feel like it makes sense.

Maybe we even should just `.tick` budget and let it refill itself when necessary? Makes sense to refill even when repair timeout is not ticking!

sentientwaffle replied:
Good point! Grid repair timeout is always ticking though, unlike `repair_budget_timeout` which gets turned off and on depending on status.

sentientwaffle replied:
:exploding_head: Ah, this was major bug! `refill(1)` is the wrong thing to do here, because `refill` resets our request tracking. i.e. it clears _all_ our tracked requests (not just 1!), which means after `refill(1)` our budget is starved until the next grid budget timeout!

sentientwaffle committed:
VSR/Replica: Refactor send_request_blocks (code review)

matklad reviewed (commented):
LGTM overall, but I want to do a more careful review tomorrow, my mind's a bit foggy today, but it should be much better after a refreshing early raise tomorrow :P

sentientwaffle committed:
VSR: refill(1) → increment

`refill(1)` is not safe, since it resets the whole budget!

sentientwaffle committed:
VSR: journal.refill max

matklad commented on src/vsr/replica.zig:
```diff
@@ -10853,67 +10853,64 @@ pub fn ReplicaType(
-            if (!self.solo()) {
-                assert(self.repair_messages_budget_grid.available >=
-                    constants.grid_repair_request_max);
-            }
```
Sorry,  I was unclear, I think we need both _this_ assert, and `assert(!self.solo());` below. The two asserts together are needed to justify that calling `decrement` is valid 


The `if (!self.solo)` one proves that we _can_ call decrement if we are not solo. The `assert(!self.solo)` proves that if we did call decrement, we are not solo.

matklad committed:
grid_blocks_missing: minor, use public API to get key/value slices

matklad committed:
grid_blocks_missing: add assert

matklad committed:
grid_blocks_missing: simplify iteration

matklad committed:
VSR/Replica: restore assert

It justifies `.decrement` below for the case where we are not solo. For
the solo case, we assert that no decrements are happening

matklad committed:
replica: assert we have enough budget for full request after refil

This justifies self.send_request_blocks(); call bellow

matklad reviewed (dismissed):
Pushed some minor cleanups&assertions, please take a look! LGTM otherwise!

sentientwaffle commented:
Your changes LGTM!
